akka-docs-dev/rst/scala/code/docs/stream/StreamPartialFlowGraphDocSpec.scala

```diff
@@ -142,4 +144,62 @@ class StreamPartialFlowGraphDocSpec extends AkkaSpec {
 
     Await.result(matSink, 300.millis) should equal(1 -> "1")
   }
+
+  "help users in providing complex ports" in {
+    object AudioPorts extends Ports {
+      val nums = Input[Int] // TODO: use I / O words or keep sinks/sources?
+      val words = Input[String]
+
+      val out = Output[String]
+    }
+
+    // building
+    val g = PartialFlowGraph(AudioPorts) { implicit b ⇒
+      import FlowGraphImplicits._
+      val merge = Merge[String]
+      // format: OFF
+      AudioPorts.nums ~> Flow[Int].map(_.toString) ~> merge
+                                  AudioPorts.words ~> merge ~> AudioPorts.out
+      // format: ON
+    }
+
+    // connecting, current style:
+    FlowGraph(g) { b ⇒
+      b.attachSource(AudioPorts.nums, Source(1 to 10))
+      b.attachSource(AudioPorts.words, Source((1 to 10).map(_.toString)))
+      b.attachSink(AudioPorts.out, Sink.ignore)
+    }.run()
+
+    // connecting, proposed style for Ports:
+    FlowGraph(g) { implicit b ⇒
+      import AudioPorts._
+
+      AudioPorts.nums(Source(1 to 10))( /* implicit */ b)
```

Comment:
I would remove this version and keep only the one that takes the Ports class as a parameter. That way the builder can take it internally as an argument, too and register them itself. Also, we can now probably unify PartialGraph and Graph, since Graph is something that takes a NoPorts as arguments and gives a RunnableGraph, anything else that takes a non-empty Ports can return a PartialGraph that gives access to those ports (API needs to be designed for that though).